jina/drivers/encode.py

```diff
@@ -1,23 +1,117 @@
 __copyright__ = "Copyright (c) 2020 Jina AI Limited. All rights reserved."
 __license__ = "Apache-2.0"
 
+from typing import Optional, Callable
+from functools import wraps
+
 from . import BaseExecutableDriver
+from ..types.sets import DocumentSet
+
+
+def encode_driver_batching_doc_set(func: Callable) -> Callable:
+    """
+    Decorator that tries to seamlessly guarantee that `_apply_all` is `EncodeDriver` is processed in batches of the desired
+    size.
+
+    """
+
+    @wraps(func)
+    def arg_wrapper(self, *args, **kwargs):
+        force_flush = True if 'force_flush' in kwargs and kwargs['force_flush'] else False
+        docs = args[0]
+        if not force_flush and \
+                self.cache_set is not None and \
+                self.cache_set.available_capacity > 0:
+            left_docs = self.cache_set.cache(docs)
+            while len(left_docs) > 0:
+                self._apply_cache()
+                left_docs = self.cache_set.cache(left_docs)
+            if self.cache_set.available_capacity == 0:
+                self._apply_cache()
```

Comment:
`func` is not used in this part, which makes it not a *real* decorator.

Reply:
yes it is used, when cache_set is full the recursivity will end up here again and call `f`

Reply:
But yes, I know itr is not a common recylcable decorator

Reply:
I will adapt it in the `_apply_all`